### ISSUE TITLE
SmartTreasuryBootstrap first review

### ISSUE THREAD

**Issue by bugduino** [AsafSilman/idle-smart-treasury#1]
Here are a couple of comments I would like to report so far:

https://github.com/AsafSilman/idle-smart-treasury/blob/b39548b87275c6d7c6161cc26f21b5f281beae8e/migrations/addresses.js#L18 this address should be changed to the one of the Timelock (https://etherscan.io/address/0xD6dABBc2b275114a2366555d6C481EF08FDC2556) instead of the GovernorAlpha . For how the governance works when we talk about 'governance ownership' we should refer to the Timelock address.

https://github.com/AsafSilman/idle-smart-treasury/blob/b39548b87275c6d7c6161cc26f21b5f281beae8e/contracts/SmartTreasuryBootstrap.sol#L96 I woudl highly suggest, in terms of gas costs, to add a parameter `bool[]` in order to skip the liquidation of some tokens, so with a keeper bot off-chain one can check if it makes sense to swap some tokens or not (eg because a per-token threshold is not reached for some tokens) and pass that parameter accordingly.

https://github.com/AsafSilman/idle-smart-treasury/blob/b39548b87275c6d7c6161cc26f21b5f281beae8e/contracts/SmartTreasuryBootstrap.sol#L110 I think here you should expect to receive at least `1`

https://github.com/AsafSilman/idle-smart-treasury/blob/b39548b87275c6d7c6161cc26f21b5f281beae8e/contracts/SmartTreasuryBootstrap.sol#L148-L157 
I'm struggling to follow the reasoning here. Could you expand the comment on this part a bit? Shouldn't be `49` as described [here](https://github.com/balancer-labs/configurable-rights-pool#example-code) in their example? I'm not sure that this `The magic number 48 is used because this is 50 - (Number of Tokens In Pool) = 48` is 48 because of the number of tokens, but I think it's rather related to the initial weights for the pool, seems that you are starting at 96/4 pool , am I missing something?

https://github.com/AsafSilman/idle-smart-treasury/blob/b39548b87275c6d7c6161cc26f21b5f281beae8e/contracts/SmartTreasuryBootstrap.sol#L256-L259 I think that this should be 'protected' with a check that ensures that the `renounce()` has been called before being enabled

On a side note I found this from the CRP [doc from balancer](https://github.com/balancer-labs/configurable-rights-pool):
```If your pool will be eligible for BAL rewards, you must apply to "redirect" the rewards to an account that can receive funds. (Without a redirect, the reward scripts would send the tokens to the CRP contract directly, where they cannot be recovered.)```
I think that this should be added somewhere

**Comment by AsafSilman** [comment]
re: addresses
I've updated the addresses in 52b017e2538503d3d26865dd81abb20c7b29ab50

re: swap()
The bootstrap contract does not need this logic, since it would be swapping all tokens, and that this is a one-time operation. This design will be implemented in the fee collector though.

re: uniswap min out amount
it is also arbitrary to expect to receive at least 1 weth for a given fee token. With the price of ethereum fluctuating as much as it is right now, it may be possible that we won't even get 1 out for some fee tokens. For this scenario, I think it is fine to not set a minimum, and just get the best available market swap. The amount of slippage we should receive is minimal.

re: weights
I had the same thought process, however, that doesn't work in some cases.
There are two rules which must be obeyed for balancer pool weights

1. The weight of a single token in a pool must be in within the range [1, 49]
2. The sum of all weights for the pool must be <= 50

Think of a situation where the value of token A is very small, and the value of token B is large
If you did this
```
W_a = val_a / tot_value * 49
W_b = val_b / tot_value * 49
```

You could have a situation where W_a is < 1. Causing an error. To fix this you could add one to W_a, and W_b

```
W_a = val_a / tot_value * 49 + 1
W_b = val_b / tot_value * 49 + 1
```

but now you see you have a problem where the sum of W_a and W_b is > 50.
You fix this by scaling the value by 48 (50-num tokens) instead of 49.
Note the +1 means that if the pool had three tokens the same problem applies, thus the scale must be adjusted, not the constant factor.

```
W_a = val_a / tot_value * 48 + 1
W_b = val_b / tot_value * 48 + 1
```

re: withdraw()
withdraw is now protected with f39105390bfb6e8c4478aec959c9368e62bd3c99

re: balancer rewards
The CRP will be created with the standard balancer factory, automatically making it 'redirect' rewards to LP, instead of the CRP contract directly. https://docs.balancer.finance/guides/crp-tutorial

**Comment by AsafSilman** [comment]
Closing this review for now, any updates and I will reopen 👍 .

**Comment by bugduino** [comment]
Re min uniswap amount: I was referring to receive at least 1 wei, if we receive 0 wei then it's better to not make the swap

Ok for everything else

**Comment by AsafSilman** [comment]
Fixed in 08361bdf9436de4f8d40033ea8dbda909e152810
